### ISSUE TITLE
when server is shut down resources should be marked down accordingly

### ISSUE THREAD

**Issue by jasoncoposky** [irods/irods#2946]
the shutdown portion of the server code should mark resources owned by the server as 'down'

**Comment by mcast** [comment]
Supposing the server is absent without clean shutdown, this won't happen.  Could there also be another way to update the status of resources, please?

**Comment by kript** [comment]
This would be quite nice if we could get it as part of 4.2?

@mcast there is already way to mark resource down as an rodsadmin, e.g.
`iadmin modresc irods-server-01-resource01 status down`
Which is something we use a fair a bit - does that fit your use case?

**Comment by kript** [comment]
> 
> 1. new status "down-auto" or regexp `m{^down(-.+)?$}i` synonymous to "down" but
>    
>    * the iRES is permitted to self-revert to "up" when it finds itself up and reaching the iCAT.

Yes, as long as 4. below is 'sticky' i.e. if a admin marks a resource down, it stays down no matter if its hosting iRES is up or down.

> 2. iRES is expected to `... modresc ... status down-auto` when it shuts down cleanly.  Maybe `down-clean` would be better?

I think there is already work on this with #2946 so 👍 

> 3. iCAT is permitted to `... modresc ... status down-auto` if it doesn't reach the resource server after ... some configurable retries.  Maybe `down-awol` would be better?

Yes, this is different from the 'down' state set by an admin. However, gets tricky once multiple IES servers; one losing routing can take out the zone, and some kind of HA quroum is expensive and complicated...
Perhaps something like the zone has some kind of raft consensus protocol for X no of servers in the zone being able to see the iRES?

> 4. humans can change `status down-auto` to `status down` if they don't want it auto-`up`ped on return to activity.
> 

Yes as above

> Just a suggestion, I'm not really the rodsadmin on a live system.

(he is a bit), but I also agree, speaking as one... I suspect this ties into the iRODs monitoring work (which I know is in request for comments stage, but I can't find to link to at the moment)

**Comment by trel** [comment]
irodsMonitor issue (#3610) and RFC (https://github.com/irods/irods_rfcs/blob/master/0002_irodsmonitor.md)

**Comment by mcast** [comment]
On Fri, Dec 21, 2018 at 04:11:49AM -0800, John Constable wrote:

> > 1. new status "down-auto" or regexp `m{^down(-.+)?$}i` synonymous
> > to "down" but
> >
> >    * the iRES is permitted to self-revert to "up" when it finds
> >      itself up and reaching the iCAT.
> 
> Yes, as long as 4. below is 'sticky' i.e. if a admin marks a
> resource down, it stays down no matter if its hosting iRES is up or
> down.

That was my intention.  I realise I am describing a state machine
which has several states, and I have suggested overloading all of
those into the one `modresc status`.


> > 3. iCAT is permitted to `... modresc ... status down-auto` if it
> > doesn't reach the resource server after ... some configurable
> > retries.  Maybe `down-awol` would be better?
> 
> Yes, this is different from the 'down' state set by an admin.

The disadvantage of `down-awol` is that the `down`ness and the
automatic state changing are two independent bits.

One reason for packing them into the single `modresc status` value is
that any change will be (closer to) atomic.


> However, gets tricky once multiple IES servers; one losing routing
> can take out the zone, and some kind of HA quroum is expensive and
> complicated...  Perhaps something like the zone has some kind of
> raft consensus protocol for X no of servers in the zone being able
> to see the iRES?

This put me off for a few weeks, but we still have a centralised
database?  The modresc state has to be recorded there anyway.

* if any IES can see the resource, state should be up

* to use a resource one must choose an IES which can reach the iRES;
  this is generally true anyway

* any IES which can't reach the database won't be committing modresc
  changes anyway 9-)


Sorry this is a bit vague,
-- 
Matthew


-- 
 The Wellcome Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.
